library/src/scala/util/boundary.scala

```diff
@@ -0,0 +1,41 @@
+package scala.util
+
+/** A boundary that can be exited by `break` calls.
+ *  `boundary` and `break` represent a unified and superior alternative for the
+ *  `scala.util.control.NonLocalReturns` and `scala.util.control.Breaks` APIs.
+ *  The main differences are:
+ *
+ *    - Unified names: `boundary` to establish a scope, `break` to leave it.
+ *      `break` can optionally return a value.
+ *    - Integration with exceptions. `break`s are logically non-fatal exceptions.
+ *      The `Break` exception class extends `RuntimeException` and is optimized so
+ *      that stack trace generation is suppressed.
+ *    - Better performance: breaks to enclosing scopes in the same method can
+ *      be rwritten to jumps.
+ */
+object boundary:
+
+  /** User code should call `break.apply` instead of throwing this exception
+   *  directly.
+   */
+  class Break[T] private[boundary](val label: Label[T], val value: T)
+  extends RuntimeException(
+    /*message*/ null, /*cause*/ null, /*enableSuppression=*/ false, /*writableStackTrace*/ false)
+
+  /** Labels are targets indicating which boundary will be exited by a `break`.
+   */
+  final class Label[-T]:
+    def break(value: T): Nothing = throw Break(this, value)
```

Comment:
We probably want this to be `private[util]`, so that the only way to do this is through `break.apply`. Otherwise, we would miss some optimizations.
```suggestion
    private[util] def break(value: T): Nothing = throw Break(this, value)
```